include/material/mohr_coulomb.tcc

```diff
@@ -0,0 +1,571 @@
+    // Define meridional eccentricity
+    double xit = 0.1;
+    // Compute Rt
+    double sqpart = 4. * (1 - et_value * et_value) * cos(theta) * cos(theta) +
+                    5. * et_value * et_value - 4. * et_value;
+    if (sqpart < 1.E-22) sqpart = 1.E-5;
+    double rt_den = 2. * (1 - et_value * et_value) * cos(theta) +
+                    (2. * et_value - 1) * std::sqrt(sqpart);
+    double rt_num = 4. * (1 - et_value * et_value) * cos(theta) * cos(theta) +
+                    (2. * et_value - 1) * (2. * et_value - 1);
+    if (fabs(rt_den) < 1.E-22) rt_den = 1.E-5;
+    double rt = rt_num / (3. * rt_den);
+    // Compute dP/dRt
+    double dp_drt = 1.5 * rho * rho * rt /
+                    std::sqrt(xit * xit * tension_cutoff_ * tension_cutoff_ +
+                              1.5 * rt * rt * rho * rho);
+    // Compute dP/dRho
+    double dp_drho = 1.5 * rho * rt * rt /
+                     std::sqrt(xit * xit * tension_cutoff_ * tension_cutoff_ +
+                               1.5 * rt * rt * rho * rho);
+    // Compute dP/dEpsilon
+    double dp_depsilon = 1. / std::sqrt(3.);
+    // Compute dRt/dThera
+    double drtden_dtheta =
+        -2. * (1 - et_value * et_value) * sin(theta) -
+        (2. * et_value - 1) * 4. * (1 - et_value * et_value) * cos(theta) *
+            sin(theta) /
+            std::sqrt(4. * (1 - et_value * et_value) * cos(theta) * cos(theta) +
+                      5. * et_value * et_value - 4. * et_value);
+    double drtnum_dtheta =
+        -8. * (1 - et_value * et_value) * cos(theta) * sin(theta);
+    double drt_dtheta = (drtnum_dtheta * rt_den - drtden_dtheta * rt_num) /
+                        (3. * rt_den * rt_den);
+    // Compute dP/dSigma
+    (*dp_dsigma) = (dp_depsilon * depsilon_dsigma) + (dp_drho * drho_dsigma) +
+                   (dp_drt * drt_dtheta * dtheta_dsigma);
+    // Compute dP/dJ
+    dp_dq = dp_drho * std::sqrt(2. / 3.);
+  }
+  // Compute dp/dsigma and dp/dj in shear yield
+  else {
+    // Compute Rmc
+    double r_mc = (3. - sin(phi)) / (6 * cos(phi));
+    // Compute deviatoric eccentricity
+    double e_val = (3. - sin(phi)) / (3. + sin(phi));
+    if (e_val <= 0.5) e_val = 0.5 + 1.E-10;
+    if (e_val > 1.) e_val = 1.;
+    // Compute Rmw
+    double sqpart = (4. * (1 - e_val * e_val) * std::pow(cos(theta), 2)) +
+                    (5 * e_val * e_val) - (4. * e_val);
+    if (sqpart < 1.E-22) sqpart = 1.E-5;
+    double m = (2. * (1 - e_val * e_val) * cos(theta)) +
+               ((2. * e_val - 1) * std::sqrt(sqpart));
+    if (fabs(m) < 1.E-22) m = 1.E-5;
+    double l = (4. * (1. - e_val * e_val) * std::pow(cos(theta), 2)) +
+               std::pow((2. * e_val - 1.), 2);
+    double r_mw = (l / m) * r_mc;
+    // Initialise meridional eccentricity
+    double xi = 0.1;
+    double omega = std::pow((xi * cohesion_peak_ * tan(psi)), 2) +
+                   std::pow((r_mw * std::sqrt(1.5) * rho), 2);
+    if (omega < 1.E-22) omega = 1.E-5;
```

Comment:
E-22 will mean 0 in double. instead, I think it's better to use `std::numeric_limits<double>::epsilon()`